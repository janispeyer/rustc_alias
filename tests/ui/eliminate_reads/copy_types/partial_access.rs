// compile-flags: -Zmir-emit-retag

#[derive(Copy, Clone)]
struct SomeTuple(i32);

#[derive(Copy, Clone)]
struct SomeStruct {
    a: i32,
    b: i32,
}

fn eliminate_reads(x: &mut SomeTuple, y: &mut SomeTuple) -> i32 {
    *x = SomeTuple(42);
    *y = SomeTuple(7);
    return x.0;
}

fn eliminate_reads_struct(x: &mut SomeStruct, y: &mut SomeStruct) -> i32 {
    *x = SomeStruct { a: 42, b: 24 };
    *y = SomeStruct { a: 17, b: 7 };
    return x.a;
}

fn main() {
    let mut x = SomeTuple(0);
    let mut y = SomeTuple(0);
    eliminate_reads(&mut x, &mut y);

    let mut a = SomeStruct { a: 0, b: 0 };
    let mut b = SomeStruct { a: 0, b: 0 };
    eliminate_reads_struct(&mut a, &mut b);
}
