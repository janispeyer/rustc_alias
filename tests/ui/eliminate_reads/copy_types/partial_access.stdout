# Analysing main
# Analysing eliminate_reads
# TopOfBorrowStack Analysis
bb0
[ 0] [_1, _2] <- Retag([fn entry] _1)
[ 1] [_1, _2] <- Retag([fn entry] _2)
[ 2] [_1, _2] <- StorageLive(_4)
[ 3] [_1, _2] <- Deinit(_4)
[ 4] [_1, _2] <- (_4.0: i32) = const 42_i32
[ 5] [_1, _2] <- (*_1) = move _4
[ 6] [_1, _2] <- StorageDead(_4)
[ 7] [_1, _2] <- StorageLive(_5)
[ 8] [_1, _2] <- Deinit(_5)
[ 9] [_1, _2] <- (_5.0: i32) = const 7_i32
[10] [_1, _2] <- (*_2) = move _5
[11] [_1, _2] <- StorageDead(_5)
[12] [_1, _2] <- _0 = ((*_1).0: i32)
[13] [] <- return

# FindImmutabilitySpans Analysis
bb0
[ 0] {} <- Retag([fn entry] _1)
[ 1] {} <- Retag([fn entry] _2)
[ 2] {} <- StorageLive(_4)
[ 3] {} <- Deinit(_4)
[ 4] {} <- (_4.0: i32) = const 42_i32
[ 5] {_1: Span(bb0[5])} <- (*_1) = move _4
[ 6] {_1: Span(bb0[5])} <- StorageDead(_4)
[ 7] {_1: Span(bb0[5])} <- StorageLive(_5)
[ 8] {_1: Span(bb0[5])} <- Deinit(_5)
[ 9] {_1: Span(bb0[5])} <- (_5.0: i32) = const 7_i32
[10] {_1: Span(bb0[5]), _2: Span(bb0[10])} <- (*_2) = move _5
[11] {_1: Span(bb0[5]), _2: Span(bb0[10])} <- StorageDead(_5)
[12] {_1: Span(bb0[5]), _2: Span(bb0[10])} <- _0 = ((*_1).0: i32)
[13] {_1: Top, _2: Top} <- return


# CFG Before
bb0
Retag([fn entry] _1)
Retag([fn entry] _2)
StorageLive(_4)
Deinit(_4)
(_4.0: i32) = const 42_i32
(*_1) = move _4
StorageDead(_4)
StorageLive(_5)
Deinit(_5)
(_5.0: i32) = const 7_i32
(*_2) = move _5
StorageDead(_5)
_0 = ((*_1).0: i32)
return

# CFG After
bb0
Retag([fn entry] _1)
Retag([fn entry] _2)
StorageLive(_4)
Deinit(_4)
(_4.0: i32) = const 42_i32
(*_1) = move _4
StorageDead(_4)
StorageLive(_5)
Deinit(_5)
(_5.0: i32) = const 7_i32
(*_2) = move _5
StorageDead(_5)
_0 = ((*_1).0: i32)
return

# Analysing eliminate_reads_struct
# TopOfBorrowStack Analysis
bb0
[ 0] [_1, _2] <- Retag([fn entry] _1)
[ 1] [_1, _2] <- Retag([fn entry] _2)
[ 2] [_1, _2] <- StorageLive(_4)
[ 3] [_1, _2] <- Deinit(_4)
[ 4] [_1, _2] <- (_4.0: i32) = const 42_i32
[ 5] [_1, _2] <- (_4.1: i32) = const 24_i32
[ 6] [_1, _2] <- (*_1) = move _4
[ 7] [_1, _2] <- StorageDead(_4)
[ 8] [_1, _2] <- StorageLive(_5)
[ 9] [_1, _2] <- Deinit(_5)
[10] [_1, _2] <- (_5.0: i32) = const 17_i32
[11] [_1, _2] <- (_5.1: i32) = const 7_i32
[12] [_1, _2] <- (*_2) = move _5
[13] [_1, _2] <- StorageDead(_5)
[14] [_1, _2] <- _0 = ((*_1).0: i32)
[15] [] <- return

# FindImmutabilitySpans Analysis
bb0
[ 0] {} <- Retag([fn entry] _1)
[ 1] {} <- Retag([fn entry] _2)
[ 2] {} <- StorageLive(_4)
[ 3] {} <- Deinit(_4)
[ 4] {} <- (_4.0: i32) = const 42_i32
[ 5] {} <- (_4.1: i32) = const 24_i32
[ 6] {_1: Span(bb0[6])} <- (*_1) = move _4
[ 7] {_1: Span(bb0[6])} <- StorageDead(_4)
[ 8] {_1: Span(bb0[6])} <- StorageLive(_5)
[ 9] {_1: Span(bb0[6])} <- Deinit(_5)
[10] {_1: Span(bb0[6])} <- (_5.0: i32) = const 17_i32
[11] {_1: Span(bb0[6])} <- (_5.1: i32) = const 7_i32
[12] {_1: Span(bb0[6]), _2: Span(bb0[12])} <- (*_2) = move _5
[13] {_1: Span(bb0[6]), _2: Span(bb0[12])} <- StorageDead(_5)
[14] {_1: Span(bb0[6]), _2: Span(bb0[12])} <- _0 = ((*_1).0: i32)
[15] {_1: Top, _2: Top} <- return


# CFG Before
bb0
Retag([fn entry] _1)
Retag([fn entry] _2)
StorageLive(_4)
Deinit(_4)
(_4.0: i32) = const 42_i32
(_4.1: i32) = const 24_i32
(*_1) = move _4
StorageDead(_4)
StorageLive(_5)
Deinit(_5)
(_5.0: i32) = const 17_i32
(_5.1: i32) = const 7_i32
(*_2) = move _5
StorageDead(_5)
_0 = ((*_1).0: i32)
return

# CFG After
bb0
Retag([fn entry] _1)
Retag([fn entry] _2)
StorageLive(_4)
Deinit(_4)
(_4.0: i32) = const 42_i32
(_4.1: i32) = const 24_i32
(*_1) = move _4
StorageDead(_4)
StorageLive(_5)
Deinit(_5)
(_5.0: i32) = const 17_i32
(_5.1: i32) = const 7_i32
(*_2) = move _5
StorageDead(_5)
_0 = ((*_1).0: i32)
return

